certifier_service/certlib/certlib_support.go

```diff
@@ -1781,8 +1793,41 @@ func GetSubjectKey(cert *x509.Certificate) *certprotos.KeyMessage {
 			fmt.Printf("GetSubjectKey: Can't internal ecc public key\n")
 			return nil
 		}
+
+		// Look for AMD VCEK cert extensions if they exist
+		oidBlSPL := "1.3.6.1.4.1.3704.1.3.1"
+		oidTeeSPL := "1.3.6.1.4.1.3704.1.3.2"
+		oidSnpSPL := "1.3.6.1.4.1.3704.1.3.3"
+		oidUcodeSPL := "1.3.6.1.4.1.3704.1.3.8"
+		snpExtExist := false
+		var blSPL, teeSPL, snpSPL, ucodeSPL uint8
+		for _, ext := range cert.Extensions {
+			if strings.Contains(ext.Id.String(), oidBlSPL) {
+				blSPL = GetVcekExtValue(ext)
+				snpExtExist = true
+			} else if strings.Contains(ext.Id.String(), oidTeeSPL) {
+				teeSPL = GetVcekExtValue(ext)
+				snpExtExist = true
+			} else if strings.Contains(ext.Id.String(), oidSnpSPL) {
+				snpSPL = GetVcekExtValue(ext)
+				snpExtExist = true
+			} else if strings.Contains(ext.Id.String(), oidUcodeSPL) {
+				ucodeSPL = GetVcekExtValue(ext)
+				snpExtExist = true
+			}
+		}
+		var tcbVer uint64
+		if snpExtExist {
```

Comment:
Did you mean GetVcekExtValue(ext) could return ^uint8(0) for one of the SPLs? I thought about whether to assert that in the if branches above. If any of the GetVcekExtValue(ext) returned failure, or any of the extensions we want are missing, I guess we can assume the VCEK cert was corrupted and the verification should fail?

Reply:
Maybe create a type like this, and set the fields to `Spl`?

```Go
type Spl struct {
	field uint8
	isSet bool
}

func (s *Spl) SetField(val uint8) {
	s.field = val
	s.isSet = true
}

func (s Spl) GetField() uint8 {
	return s.field
}

func (s Spl) IsSet() bool {
	return s.isSet
}
```

Then do this:

```suggestion
		if snpExtExist && blSPL.IsSet() && teeSPL.IsSet() && snpSPL.IsSet() && ucodeSPL.IsSet() {
		...
```

Reply:
I could do this, or equivalently, check for whether GetVcekExtValue(ext) returns 0xFF for any of the if statements and bail if any of them does. In this case, tcbVer will be set to 0xFFFFFFFF, which will skip the check (we currently skip the check if the extensions do not exist in the VCEK cert. This makes the simulator cert and attestation verification pass).

Alternatively, do you feel we should fail the verification instead of skipping it in this case?

Reply:
> Did you mean GetVcekExtValue(ext) could return ^uint8(0) for one of the SPLs?

I was thinking more along the lines of if the extension isn't found, the default value in Go is zero, which is technically a valid value.

Reply:
Understood. My current thought is like this:

1. All values from 0x0 - 0xFF are valid extension values for the various SPLs
2. GetVcekExtValue(ext) will return {bool res, uint8 value} pair
3. If any of the extensions are missing, we will bail and set tcbVer to 0xFFFFFFFF
4. If tcbVer is 0xFFFFFFFF, then tcb version check during the report verification will be skipped

Now, the questions are: 1. Is bullet 3 assumption reasonable? And 2. Is bullet 4 logic reasonable?

Additionally, I do also have a question about ucodeSPL. In the cert, it seems to be a 16 bit value. But when the specification talks about TCB_VERSION, it assumes ucodeSPL is 8 bit. In the code above, I assumed ucodeSPL is 8 bit instead. Is this OK?